core/src/main/java/io/grpc/internal/AbstractServerStream.java

```diff
@@ -31,256 +31,207 @@
 
 package io.grpc.internal;
 
-import static com.google.common.base.Preconditions.checkNotNull;
-
 import com.google.common.base.Preconditions;
 
 import io.grpc.Attributes;
 import io.grpc.Metadata;
 import io.grpc.Status;
 
-import java.io.InputStream;
-import java.util.logging.Level;
-import java.util.logging.Logger;
+import javax.annotation.Nullable;
 
 /**
- * Abstract base class for {@link ServerStream} implementations.
- *
- * @param <IdT> the type of the stream identifier
+ * Abstract base class for {@link ServerStream} implementations. Extending classes only need to
+ * implement {@link #transportState()} and {@link #abstractServerStreamSink()}.
  */
-public abstract class AbstractServerStream<IdT> extends AbstractStream<IdT>
-    implements ServerStream {
-  private static final Logger log = Logger.getLogger(AbstractServerStream.class.getName());
-
-  /** Whether listener.closed() has been called. */
-  private boolean listenerClosed;
-  private ServerStreamListener listener;
-
-  private boolean headersSent = false;
+public abstract class AbstractServerStream extends AbstractStream2
+    implements ServerStream, MessageFramer.Sink {
   /**
-   * Whether the stream was closed gracefully by the application (vs. a transport-level failure).
+   * Only called from application thread. A separate interface just to avoid name
+   * collisions/confusion.
    */
-  private boolean gracefulClose;
-  /** Saved trailers from close() that need to be sent once the framer has sent all messages. */
-  private Metadata stashedTrailers;
+  protected interface Sink {
+    /**
+     * Sends response headers to the remote end points.
+     *
+     * @param headers the headers to be sent to client.
+     */
+    void writeHeaders(Metadata headers);
 
-  protected AbstractServerStream(WritableBufferAllocator bufferAllocator,
-                                 int maxMessageSize) {
-    super(bufferAllocator, maxMessageSize);
+    /**
+     * Sends an outbound frame to the remote end point.
+     *
+     * @param frame a buffer containing the chunk of data to be sent.
+     * @param flush {@code true} if more data may not be arriving soon
+     */
+    void writeFrame(@Nullable WritableBuffer frame, boolean flush);
+
+    /**
+     * Sends trailers to the remote end point. This call implies end of stream.
+     *
+     * @param trailers metadata to be sent to end point
+     * @param headersSent {@code true} if response headers have already been sent.
```

Comment:
Why does the application need to keep track of whether the initial headers have been sent?  The idea of having a separate methods from writeHeaders is the right API abstraction.  

Having this extra boolean seems wrong.  If the application has to keep track of this anyway, the API might as well be:

writeHeaders(Metadata metadata, boolean isTrailers);

and just have one method.